File: solutions/misc/frequently-asked-questions/bfs-problems.js
/**
 * BFS — 7 Core Problems
 *
 * Template:
 *   queue = [start]; visited = new Set([start])
 *   while queue:
 *     node = queue.shift()
 *     for neighbor of graph[node]:
 *       if not visited: visited.add; queue.push
 *
 * Problems:
 *   LC 102 — Binary Tree Level Order Traversal (Medium)
 *   LC 200 — Number of Islands (Medium)
 *   LC 994 — Rotting Oranges (Medium)
 *   LC 542 — 01 Matrix (Medium)
 *   LC 127 — Word Ladder (Hard) — see word-ladder.js
 *   LC 286 — Walls and Gates (Medium)
 *   LC 815 — Bus Routes (Hard)
 */

// ═══════════════════════════════════════════════════════
// LC 102 — Binary Tree Level Order Traversal
// ═══════════════════════════════════════════════════════
function levelOrder(root) {
  if (!root) return [];
  const res = [];
  const queue = [root];
  while (queue.length) {
    const levelSize = queue.length;
    const level = [];
    for (let i = 0; i < levelSize; i++) {
      const node = queue.shift();
      level.push(node.val);
      if (node.left)  queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    res.push(level);
  }
  return res;
}

// ═══════════════════════════════════════════════════════
// LC 200 — Number of Islands
// ═══════════════════════════════════════════════════════
function numIslands(grid) {
  const rows = grid.length, cols = grid[0].length;
  const dirs = [[0,1],[0,-1],[1,0],[-1,0]];
  let count = 0;

  function bfs(r, c) {
    const queue = [[r, c]];
    grid[r][c] = "0";
    while (queue.length) {
      const [row, col] = queue.shift();
      for (const [dr, dc] of dirs) {
        const nr = row + dr, nc = col + dc;
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] === "1") {
          queue.push([nr, nc]);
          grid[nr][nc] = "0";
        }
      }
    }
  }

  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++)
      if (grid[r][c] === "1") { count++; bfs(r, c); }

  return count;
}

// ═══════════════════════════════════════════════════════
// LC 994 — Rotting Oranges
// ═══════════════════════════════════════════════════════
function orangesRotting(grid) {
  const rows = grid.length, cols = grid[0].length;
  const dirs = [[0,1],[0,-1],[1,0],[-1,0]];
  const queue = [];
  let fresh = 0;

  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++) {
      if (grid[r][c] === 2) queue.push([r, c, 0]);
      if (grid[r][c] === 1) fresh++;
    }

  let maxTime = 0;
  while (queue.length) {
    const [r, c, time] = queue.shift();
    for (const [dr, dc] of dirs) {
      const nr = r + dr, nc = c + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] === 1) {
        grid[nr][nc] = 2;
        fresh--;
        maxTime = Math.max(maxTime, time + 1);
        queue.push([nr, nc, time + 1]);
      }
    }
  }
  return fresh === 0 ? maxTime : -1;
}

// ═══════════════════════════════════════════════════════
// LC 542 — 01 Matrix (distance to nearest 0)
// ═══════════════════════════════════════════════════════
function updateMatrix(mat) {
  const rows = mat.length, cols = mat[0].length;
  const dist = Array.from({ length: rows }, () => new Array(cols).fill(Infinity));
  const queue = [];
  const dirs = [[0,1],[0,-1],[1,0],[-1,0]];

  for (let r = 0; r < rows; r++)
    for (let c = 0; c < cols; c++)
      if (mat[r][c] === 0) { dist[r][c] = 0; queue.push([r, c]); }

  while (queue.length) {
    const [r, c] = queue.shift();
    for (const [dr, dc] of dirs) {
      const nr = r + dr, nc = c + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && dist[nr][nc] > dist[r][c] + 1) {
        dist[nr][nc] = dist[r][c] + 1;
        queue.push([nr, nc]);
      }
    }
  }
  return dist;
}

// ═══════════════════════════════════════════════════════
// LC 815 — Bus Routes (minimum number of buses to reach target)
// ═══════════════════════════════════════════════════════
function numBusesToDestination(routes, source, target) {
  if (source === target) return 0;

  // Build stop → list of route indices
  const stopToRoutes = new Map();
  for (let i = 0; i < routes.length; i++)
    for (const stop of routes[i]) {
      if (!stopToRoutes.has(stop)) stopToRoutes.set(stop, []);
      stopToRoutes.get(stop).push(i);
    }

  const visitedStops  = new Set([source]);
  const visitedRoutes = new Set();
  const queue = [[source, 0]]; // [stop, buses_taken]

  while (queue.length) {
    const [stop, buses] = queue.shift();
    for (const routeIdx of (stopToRoutes.get(stop) || [])) {
      if (visitedRoutes.has(routeIdx)) continue;
      visitedRoutes.add(routeIdx);
      for (const nextStop of routes[routeIdx]) {
        if (nextStop === target) return buses + 1;
        if (!visitedStops.has(nextStop)) {
          visitedStops.add(nextStop);
          queue.push([nextStop, buses + 1]);
        }
      }
    }
  }
  return -1;
}

// ─── Tests ───────────────────────────────────────────────────────────────────
const grid1 = [["1","1","1"],["0","1","0"],["1","1","1"]];
console.log("Islands:", numIslands(grid1)); // 1

const rotten = [[2,1,1],[1,1,0],[0,1,1]];
console.log("Rotting:", orangesRotting(rotten)); // 4

console.log("Bus routes:", numBusesToDestination([[1,2,7],[3,6,7]], 1, 6)); // 2

module.exports = { levelOrder, numIslands, orangesRotting, updateMatrix, numBusesToDestination };

File: solutions/misc/frequently-asked-questions/bfs-problems.py
"""
BFS — 7 Core Problems (Python)

LC 102 — Binary Tree Level Order Traversal
LC 200 — Number of Islands
LC 994 — Rotting Oranges
LC 542 — 01 Matrix
LC 127 — Word Ladder (see word-ladder.py)
LC 286 — Walls and Gates
LC 815 — Bus Routes
"""

from collections import deque, defaultdict
from typing import List, Optional


# ── TreeNode helper ───────────────────────────────────────────────────────────
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


# LC 102 — Level Order Traversal
def level_order(root: Optional[TreeNode]) -> List[List[int]]:
    """BFS level order traversal of binary tree."""
    if not root: return []
    res, queue = [], deque([root])
    while queue:
        level = []
        # Process all nodes at current level (queue size = level size)
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:  queue.append(node.left)
            if node.right: queue.append(node.right)
        res.append(level)
    return res


# LC 200 — Number of Islands
def num_islands(grid: List[List[str]]) -> int:
    """Count connected components of '1's in 2D grid using BFS."""
    rows, cols = len(grid), len(grid[0])
    dirs = [(0,1),(0,-1),(1,0),(-1,0)]  # 4-directional movement
    count = 0

    def bfs(r, c):
        """Mark all connected land cells as visited (change to '0')."""
        queue = deque([(r, c)])
        grid[r][c] = "0"  # Mark as visited
        while queue:
            row, col = queue.popleft()
            for dr, dc in dirs:
                nr, nc = row+dr, col+dc
                # If neighbor is valid land cell, add to queue and mark visited
                if 0<=nr<rows and 0<=nc<cols and grid[nr][nc]=="1":
                    queue.append((nr, nc)); grid[nr][nc]="0"

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == "1": count += 1; bfs(r, c)
    return count


# LC 994 — Rotting Oranges
def oranges_rotting(grid: List[List[int]]) -> int:
    """Calculate minimum minutes for all oranges to rot using multi-source BFS."""
    rows, cols = len(grid), len(grid[0])
    dirs = [(0,1),(0,-1),(1,0),(-1,0)]
    queue, fresh = deque(), 0
    # Initialize queue with all rotten oranges (2) and count fresh oranges (1)
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == 2: queue.append((r, c, 0))  # (row, col, time)
            elif grid[r][c] == 1: fresh += 1
    max_time = 0
    # BFS from all rotten oranges simultaneously
    while queue:
        r, c, t = queue.popleft()
        for dr, dc in dirs:
            nr, nc = r+dr, c+dc
            # If neighbor is fresh orange, rot it and add to queue
            if 0<=nr<rows and 0<=nc<cols and grid[nr][nc]==1:
                grid[nr][nc]=2; fresh-=1; max_time=max(max_time,t+1); queue.append((nr,nc,t+1))
    return max_time if fresh==0 else -1  # -1 if some oranges never rot


# LC 542 — 01 Matrix
def update_matrix(mat: List[List[int]]) -> List[List[int]]:
    """Compute distance to nearest 0 for each cell using multi-source BFS."""
    rows, cols = len(mat), len(mat[0])
    dist = [[float("inf")]*cols for _ in range(rows)]
    queue = deque()
    dirs = [(0,1),(0,-1),(1,0),(-1,0)]
    # Initialize queue with all 0 cells (distance 0)
    for r in range(rows):
        for c in range(cols):
            if mat[r][c]==0: dist[r][c]=0; queue.append((r,c))
    # BFS to propagate distances
    while queue:
        r, c = queue.popleft()
        for dr, dc in dirs:
            nr, nc = r+dr, c+dc
            # If neighbor can be reached with shorter distance, update and add to queue
            if 0<=nr<rows and 0<=nc<cols and dist[nr][nc]>dist[r][c]+1:
                dist[nr][nc]=dist[r][c]+1; queue.append((nr,nc))
    return dist


# LC 815 — Bus Routes
def num_buses_to_destination(routes: List[List[int]], source: int, target: int) -> int:
    """Find minimum number of buses to reach target from source using BFS on route graph."""
    if source == target: return 0
    # Build map: stop -> set of route indices that include this stop
    stop_to_routes: defaultdict = defaultdict(set)
    for i, route in enumerate(routes):
        for stop in route: stop_to_routes[stop].add(i)

    visited_stops = {source}
    visited_routes: set = set()
    queue = deque([(source, 0)])  # (stop, number_of_buses_taken)

    while queue:
        stop, buses = queue.popleft()
        # For each route that passes through current stop
        for route_idx in stop_to_routes[stop]:
            if route_idx in visited_routes: continue
            visited_routes.add(route_idx)
            # Explore all stops on this route
            for next_stop in routes[route_idx]:
                if next_stop == target: return buses + 1
                if next_stop not in visited_stops:
                    visited_stops.add(next_stop); queue.append((next_stop, buses+1))
    return -1


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    grid = [["1","1","1"],["0","1","0"],["1","1","1"]]
    print("Islands:", num_islands(grid))  # 1

    rotten = [[2,1,1],[1,1,0],[0,1,1]]
    print("Rotting:", oranges_rotting(rotten))  # 4

    print("Bus routes:", num_buses_to_destination([[1,2,7],[3,6,7]], 1, 6))  # 2
